lib/model/Product.dart

```diff
@@ -305,45 +305,52 @@ class Product extends JsonObject {
     for (final String key in json.keys) {
       if (key.contains('debug')) {
         continue;
-      } else if (key.startsWith('product_name_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'product_name_');
-        if (lang != OpenFoodFactsLanguage.UNDEFINED) {
-          result.productNameInLanguages ??= {};
-          result.productNameInLanguages![lang] = json[key];
-        }
-      } else if (key.startsWith('categories_tags_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'categories_tags_');
-        final values = _jsonValueToList(json[key]);
-        if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
-          result.categoriesTagsInLanguages ??= {};
-          result.categoriesTagsInLanguages![lang] = values;
-        }
-      } else if (key.startsWith('ingredients_tags_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'ingredients_tags_');
-        final values = _jsonValueToList(json[key]);
-        if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
-          result.ingredientsTagsInLanguages ??= {};
-          result.ingredientsTagsInLanguages![lang] = values;
-        }
-      } else if (key.startsWith('labels_tags_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'labels_tags_');
-        final values = _jsonValueToList(json[key]);
-        if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
-          result.labelsTagsInLanguages ??= {};
-          result.labelsTagsInLanguages![lang] = values;
-        }
-      } else if (key.startsWith('countries_tags_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'countries_tags_');
-        final values = _jsonValueToList(json[key]);
-        if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
-          result.countriesTagsInLanguages ??= {};
-          result.countriesTagsInLanguages![lang] = values;
-        }
-      } else if (key.startsWith('ingredients_text_')) {
-        OpenFoodFactsLanguage lang = _langFrom(key, 'ingredients_text_');
-        if (lang != OpenFoodFactsLanguage.UNDEFINED) {
-          result.ingredientsTextInLanguages ??= {};
-          result.ingredientsTextInLanguages![lang] = json[key];
+      } else {
+        // If the key ends with _[2 characters], it might be a language specific
+        // field.
+        RegExp regexp = RegExp(r'_(\w\w)$');
+        if (regexp.hasMatch((key))) {
```

Comment:
I'm not fluent in regexp, therefore I would have coded that differently. There are explicit comments, fair enough.

I still think that you could get rid of your regexp just by saying:
* I compute the language
* if the language is unknown, I do nothing
* else I do something, which includes using the problematic `_jsonValueToList`

Probably would be easier to read. What do you think of that?

Reply:
Ah, I use regexps a lot (Perl is my favorite language) so it didn't strike me as bad, but there are other ways certainly.

I'm not sure I understand your suggestion, you want to move the check in each of the conditions?

```
else if (key.startsWith('categories_tags_')) {
            OpenFoodFactsLanguage lang = _langFrom(key, 'categories_tags_');
            final values = _jsonValueToList(json[key]);
            if (lang != OpenFoodFactsLanguage.UNDEFINED && values != null) {
              result.categoriesTagsInLanguages ??= {};
              result.categoriesTagsInLanguages![lang] = values;
            }
          }
```

So something like:

```
else if (key.startsWith('categories_tags_')) {
            OpenFoodFactsLanguage lang = _langFrom(key, 'categories_tags_');
             if (lang != OpenFoodFactsLanguage.UNDEFINED) {
               final values = _jsonValueToList(json[key]);
                if (values != null) {
                 result.categoriesTagsInLanguages ??= {};
                 result.categoriesTagsInLanguages![lang] = values;
               }
            }
          }
```

? (repeated for labels_tags, countries_tags etc.)

It's confusing I think: it would sound as if we somehow get something like  category_tags_zx and zx is not recognized as a language.

Here what we want is to make sure not to match categories_tags_in_languages. I guess we could explicitly exclude it:

else if (key.startsWith('categories_tags_') && key != 'categories_tags_in_languages') {

Then it would be more clear why we have this condition.

Reply:
Correct, you've understood my point: as anyway we're checking if the language is "defined" (I did not invent it, it's already coded), we can skip the cryptic regexp (at least cryptic enough that you needed to comment it) by checking the language as early as possible.

I agree with the last part of your comment too: it could be more readable to explicitly say "it's not just about wrong language codes, it's also about `_in_languages`".
That would be something like that:
```dart
else if (key.startsWith('categories_tags_')) {
  if (key != 'categories_tags_in_languages') { // more explicit indeed
    final OpenFoodFactsLanguage lang = _langFrom(key, 'categories_tags_');
    if (lang != OpenFoodFactsLanguage.UNDEFINED) {
      final values = _jsonValueToList(json[key]);
      if (values != null) {
        result.categoriesTagsInLanguages ??= {};
        result.categoriesTagsInLanguages![lang] = values;
      }
    }
  }
}
```
or perhaps better
```dart
if (key.contains('debug')) {
   continue;
}
if (key.endsWith('_in_languages')) {
   continue;
}
if (key.startsWith('categories_tags_')) {
  final OpenFoodFactsLanguage lang = _langFrom(key, 'categories_tags_');
  if (lang != OpenFoodFactsLanguage.UNDEFINED) {
    final values = _jsonValueToList(json[key]); 
    if (values != null) {
      result.categoriesTagsInLanguages ??= {};
      result.categoriesTagsInLanguages![lang] = values;
  }
}
```

PS: for the record I hate Perl :)

Reply:
Sounds good, I replaced the regexp with:

```
if (key.endsWith('_in_languages')) {
   continue;
}
```

and added a few comments to explain why we need to do this.